daceml/onnx/op_implementations/fpga_implementations.py

```diff
@@ -886,6 +886,1132 @@ def make_compute(sdfg, state, vec_width=1):
+                                  dst_conn="a_in")
+            state.add_memlet_path(buffer_a_tasklet,
+                                  A_reg,
+                                  memlet=dace.Memlet("A_reg[0]", dynamic=True),
+                                  src_conn="a_reg")
+
+            # Feed B
+            sdfg.add_array("B_reg",
+                           shape=[1],
+                           dtype=vec_type,
+                           transient=True,
+                           storage=dace.dtypes.StorageType.FPGA_Local)
+            B_reg = state.add_access("B_reg")
+            buffer_b_tasklet = state.add_tasklet(
+                "buffer_b", {"b_in"}, {"b_reg_out"}, f"""\
+if  m>={L} and not {entry_pipeline.pipeline.drain_condition()}:
+    b_reg_out = b_in""")
+
+            state.add_memlet_path(B_pipe_in,
+                                  entry_pipeline,
+                                  buffer_b_tasklet,
+                                  memlet=dace.Memlet("B_pipe[p]",
+                                                     dynamic=True),
+                                  dst_conn="b_in")
+            state.add_memlet_path(buffer_b_tasklet,
+                                  B_reg,
+                                  memlet=dace.Memlet("B_reg[0]", dynamic=True),
+                                  src_conn="b_reg_out")
+
+            # Compute, Forward B, and Drain
+            compute_tasklet = state.add_tasklet(
+                "compute_and_drain",
+                {"a_in", "b_in", "c_in", "forward_in", "c_init_data"},
+                {"b_out", "c_out", "c_pipe_out"}, f"""\
+result = c_in
+if m >= {int(L)} and not {entry_pipeline.pipeline.drain_condition()}:
+    c_prev = c_init_data if k == 0 else c_in
+    result =  c_prev + a_in * b_in
+    c_out = result
+    if p < {P} - 1:
+        b_out = b_in
+# Drain
+# when we have to drain:
+# - if we are working on second assigned row or second tile and we have something to drain
+# - if k = K-1 and m>=L: each PE has just finished to compute something
+# - if we are in the draining phase
+# How: 
+# - if k = K-1 and m>=L: then the PE drains its own result
+#-  otherwise, if k_drain<p forward data coming from previous PEs (this could happens also in the drain phase)
+if((b > 0 or n0 > 0 or tm > 0)  and k_drain <p and m_drain <{int(T / vec_width)}) or  (k=={K}-1 and m>= {int(L)}) or ({entry_pipeline.pipeline.drain_condition()} and k_drain < p): # modification to standard GEMM, also consider b
+    c_pipe_out = result if (p==0 or (k_drain=={K}-1 and not {entry_pipeline.pipeline.drain_condition()})) else forward_in
+# adjust draining iterators
+if not {entry_pipeline.pipeline.drain_condition()}:
+    if m_drain >= {int(L)} +  {int(T / vec_width)} -1:
+        m_drain = 0
+        if k_drain >= {K} - 1:
+            k_drain = 0
+        else:
+            k_drain = k_drain +1
+    else:
+        m_drain = m_drain + 1
```

Comment:
Do you think this is a missing feature in the DaCe pipeline abstraction?

Reply:
Do you mean the safe delay `L` offset check or in general the draining iterators? It is hard for me to tell, as I haven't seen to many systolic array applications. So I can't really say how something like this would look like when one tries to generalise it to other applications. So I mean it is probably a pattern that one sees often, that we need to keep track of the iterations during draining, but as said, might be difficult to generalise such that it becomes useful.
Also it probably occurs more with systolic arrays, than with a "normal" pipeline, but this might just be me not having seen enough examples. If you also think this is the case, it might be more suitable to try and create a systolic array abstraction, but I from what I have seen/heard so far, this would be worth a paper on its own, if it can be made to work.

Reply:
The draining iterators I guess. If it's dependent on the position in the systolic array, then I guess it doesn't really generalize though.